(Feature) New pagination view

gabitoesmiapodo opened the pull request:
Closes #1893

## Motivation

Pagination styles need to be updated according to this design:

![Screen Shot 2019-05-10 at 16 33 20](https://user-images.githubusercontent.com/4015436/57552147-5a814300-7341-11e9-9ee5-e22d16ff6a79.png)

This is obviously missing some logic which the design shows but it's absent from the current implementation:

- `First` and `Last` buttons.
- `Page 1 of X` needs to calculate the value for `X`.
- `Show XXXX Records`, amount of items to show should be implemented too (this can be hidden using a parameter).

Sections including the pagination component:

- Blocks / Uncles / Forked.
- Transactions Validated / Pending.
- Accounts.

## Screenshots

(Colors will obviously be different depending on your current theme)

![Screen Shot 2019-05-10 at 16 30 21](https://user-images.githubusercontent.com/4015436/57552467-615c8580-7342-11e9-9f0d-ba186359672b.png)
![Screen Shot 2019-05-10 at 16 30 30](https://user-images.githubusercontent.com/4015436/57552468-615c8580-7342-11e9-869d-b174c3bbf61e.png)
![Screen Shot 2019-05-10 at 16 30 53](https://user-images.githubusercontent.com/4015436/57552469-615c8580-7342-11e9-8217-b16b41f9b8ac.png)

## Changelog

### Enhancements

- Pagination styles were updated.

## Checklist for your PR

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [x] I checked whether I should update the docs and did so if necessary

ayrat555 commented:
`Page 1 of undefined` is it ok?

Also, if there are no blocks should we show navigation buttons?

ayrat555 commented on apps/block_scout_web/lib/block_scout_web/templates/common_components/_pagination_container.html.eex:
```diff
@@ -0,0 +1,52 @@
+                </svg>
+            </a>
+        </li>
+        <li class="page-item"><a class="page-link no-hover" href><%= gettext "Page" %> <%= assigns[:cur_page_number] || "" %> <%= gettext "of" %> <%= assigns[:total_pages_number] || "undefined" %></a></li>
```
maybe we should show only current page if there is no `total_pages_number` instead of `of undefined`

gabitoesmiapodo replied:
Done.

I think everything else will be addressed in https://github.com/poanetwork/blockscout/issues/1930

vbaranov commented:
@gabitoesmiapodo 

In addition, we have pagination buttons for transactions list on address details page like this: https://blockscout.com/poa/core/address/0xc76815f3815614f10b78dfe94185e86d9f5250c5/transactions. We should also use new pagination here.

vbaranov commented:
@gabitoesmiapodo 

plus, we should have the same pagination for transactions in the block details page. For example, here https://blockscout.com/eth/mainnet/blocks/7751624/transactions

gabitoesmiapodo commented:
@vbaranov 

> As you correctly mentioned there are some new buttons and missing logic for new pagination. And I created a separate issue to fill those missing parts of the logic #1930. In the meantime, the logic for left/right pagination buttons exists in the current pagination, but it wasn't ported to the new pagination buttons. Could you return it? Because for now, pagination is rather a mockup.

I didn't get that part, could you please clarify?

Working on fixing everything else.

vbaranov commented:
@gabitoesmiapodo 

>>As you correctly mentioned there are some new buttons and missing logic for new pagination. And I created a separate issue to fill those missing parts of the logic #1930. In the meantime, the logic for left/right pagination buttons exists in the current pagination, but it wasn't ported to the new pagination buttons. Could you return it? Because for now, pagination is rather a mockup.

>I didn't get that part, could you please clarify?

Sure, I mean, in the current pagination (that we would like to update) we already have implemented left|right buttons and logic for them:

![Screenshot 2019-05-13 at 14 08 46](https://user-images.githubusercontent.com/4341812/57617202-e46f1d00-7588-11e9-8da6-3e2868f664c5.png)

And that logic has been removed from the new left|right buttons (that are from this PR). Could we return that logic to those buttons before merging?

gabitoesmiapodo commented:
@vbaranov 

Oh, sure.

I thought that I've left that working, but it seems it's not.

vbaranov commented:
@gabitoesmiapodo 

in addition, the same pagination should be on the address details page for
- [coin balances](https://blockscout.com/eth/mainnet/address/0x742d35cc6634c0532925a3b844bc454e4438f44e/coin_balances)
- [tokens](https://blockscout.com/eth/mainnet/address/0x742d35cc6634c0532925a3b844bc454e4438f44e/tokens)
- internal transactions

gabitoesmiapodo commented:
@vbaranov 

> The old pagination messages like this Showing 10 Validated Transactions or Showing 50 addresses of 0 total addresses with a balance (page 1) should be removed as it is implemented in the new pagination view.

Removed all text.

> Could we display pagination buttons in the mobile view aligned to the center of the screen (see the screenshot below)?

Done.

> As you correctly mentioned there are some new buttons and missing logic for new pagination. And I created a separate issue to fill those missing parts of the logic #1930. In the meantime, the logic for left/right pagination buttons exists in the current pagination, but it wasn't ported to the new pagination buttons. Could you return it? Because for now, pagination is rather a mockup.

I've only found that combination `Back / Next` on the _Accounts_ section. Seems to be working now, but it is disabled if there's no back or next page.

> in addition, the same pagination should be on the address details page for

I've replaced the pagination in every other section too, but in those cases there was only an `Older` or `Next` button. In that case, it's going to be replaced by the `>` button until its implementation is updated.

gabitoesmiapodo commented:
@vbaranov 

> Could we change the paddings for...

Sure, done. It should look like this:

![Screen Shot 2019-05-14 at 08 50 23](https://user-images.githubusercontent.com/4015436/57695714-5bf28a00-7625-11e9-9499-c9062aa9572e.png)


> I see now that the pagination buttons work for Accounts page, but could we also return the logic to the > button, which is instead of Older or Next button?

Not sure if I follow.

For the "Accounts" section there were only `Back` and `Next` buttons. Those were converted to `<` and `>` now.

For the rest of the pages I only found one `Older` (or `Next`) isolated button, which is now the `>` button.

These buttons are disabled when there's nothing to paginate (`@prev_page_path` / `@next_page_path` are null, etc.), but they could be hidden instead.

So, the functionality should be there (unless I screwed something up, which could be the case).

Maybe I inadvertently removed some other functionality?

vbaranov commented:
@gabitoesmiapodo 

I am trying to open blocks page, transactions page or other pages with pagination from your PR. On those pages, we have currently in the prod `Older` or `Next` buttons, and they work fine in the prod. When I open those pages from this PR, `>` button is always disabled. Please, try it locally.

gabitoesmiapodo commented:
@vbaranov 

Got it now, thank you.

It should work as before now, warts and all. Notice that the button will be disabled once there are no more pages, and there's no current way to go back.

![next](https://user-images.githubusercontent.com/4015436/57701761-045b1b00-7633-11e9-9b39-7ba91b6e649a.gif)

vbaranov commented:
@gabitoesmiapodo yes, seems, pagination buttons work now. The last thing I forgot to mention in the previous review is:

![Screenshot 2019-05-14 at 12 43 05](https://user-images.githubusercontent.com/4341812/57708377-e234bd80-7671-11e9-8c12-d49776bb632b.png)

we should move the top pagination below the chart because the cart is not responsive on changing pages. It shows the whole data.

gabitoesmiapodo commented:
@vbaranov 

Done.
